torch/nn/modules/module.py

```diff
@@ -1509,78 +1526,118 @@ def _call_impl(self, *args, **kwargs):
+                                args, kwargs = args_kwargs_result
+                            else:
+                                raise RuntimeError(
+                                    "forward pre-hook must return None or a tuple "
+                                    f"of (new_args, new_kwargs), but got {args_kwargs_result}."
+                                )
+                    else:
+                        args_result = hook(self, args)
+                        if args_result is not None:
+                            if not isinstance(args_result, tuple):
+                                args_result = (args_result,)
+                            args = args_result
+
+            bw_hook = None
+            if full_backward_hooks or backward_pre_hooks:
+                bw_hook = hooks.BackwardHook(self, full_backward_hooks, backward_pre_hooks)
+                args = bw_hook.setup_input_hook(args)
+
+            result = forward_call(*args, **kwargs)
+            if _global_forward_hooks or self._forward_hooks:
+                for hook_id, hook in (
+                    *_global_forward_hooks.items(),
+                    *self._forward_hooks.items(),
+                ):
+                    # mark that always called hook is run
+                    if hook_id in self._forward_hooks_always_called or hook_id in _global_forward_hooks_always_called:
+                        called_always_called_hooks.add(hook_id)
+
+                    if hook_id in self._forward_hooks_with_kwargs:
+                        hook_result = hook(self, args, kwargs, result)
+                    else:
+                        hook_result = hook(self, args, result)
+
+                    if hook_result is not None:
+                        result = hook_result
+
+            if bw_hook:
+                if not isinstance(result, (torch.Tensor, tuple)):
+                    warnings.warn("For backward hooks to be called,"
+                                  " module output should be a Tensor or a tuple of Tensors"
+                                  f" but received {type(result)}")
+                result = bw_hook.setup_output_hook(result)
+
+            # Handle the non-full backward hooks
+            if non_full_backward_hooks:
+                var = result
+                while not isinstance(var, torch.Tensor):
+                    if isinstance(var, dict):
+                        var = next((v for v in var.values() if isinstance(v, torch.Tensor)))
+                    else:
+                        var = var[0]
+                grad_fn = var.grad_fn
+                if grad_fn is not None:
+                    for hook in non_full_backward_hooks:
+                        grad_fn.register_hook(_WrappedHook(hook, self))
+                    self._maybe_warn_non_full_backward_hook(args, result, grad_fn)
+
+            return result
+
+        except Exception:
```

Comment:
seems like dynamo does not trace `finally:` blocks 

```
2023-06-28T03:46:21.5933362Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/_dynamo/symbolic_convert.py", line 2229, in inline_call_
2023-06-28T03:46:21.5933502Z     tracer.run()
2023-06-28T03:46:21.5933915Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/_dynamo/symbolic_convert.py", line 708, in run
2023-06-28T03:46:21.5934012Z     and self.step()
2023-06-28T03:46:21.5934429Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/_dynamo/symbolic_convert.py", line 664, in step
2023-06-28T03:46:21.5934592Z     unimplemented(f"missing: {inst.opname}")
2023-06-28T03:46:21.5934998Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/_dynamo/exc.py", line 140, in unimplemented
2023-06-28T03:46:21.5935128Z     raise Unsupported(msg)
2023-06-28T03:46:21.5935318Z torch._dynamo.exc.Unsupported: missing: CALL_FINALLY
```
It runs into this part 
https://github.com/pytorch/pytorch/blob/044a8e3305bdff28780cdab757b859abf2fc76d9/torch/_dynamo/exc.py#L139-L141

so I changed this to 
```
try:
   ...
except:
   ...
   raise
```

Reply:
cc @jansel any context on this?

Reply:
Not sure what the question is here, unimplemented will fall back to eager since `CALL_FINALLY` isn't implemented.

Reply:
@jansel Got it, I assume fall back to eager constitute a graph break? To check my understanding, I guess this means that we shouldn't use `finally:` here since this change to the Module._call_impl will cause graph breaks on all modules? So using `except + raise` (which seems like it doesn't trigger graph break)  should be better here is that correct?

Reply:
Either that or we add support for the CALL_FINALLY bytecode.